### PR TITLE
add prod op

### PR DESCRIPTION
**Pull request by QiangX-man**
- 对齐pytorch.prod算子
- 延指定axis，求对应行的所有元素之积

图片

图片

### CODE REVIEW

File: python/oneflow/nn/modules/reduce_ops.py [hunk 1]
```diff
@@ -161,6 +161,44 @@ def _max(input, dim=None, keepdim=False):
     return Max(dim, keepdim)(input)
 
 
+def prod_op(input, dim=None):
```
Comment (BBuf) [inline python/oneflow/nn/modules/reduce_ops.py:164]: 还有一个keepdim参数

File: python/oneflow/test/modules/test_prod.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    def test_reduce_prod_without_dim(test_case):
+        device = random_device()
+
+        ndim = random(1, 5).to(int)
+        x = random_pytorch_tensor(ndim=ndim).to(device)
+        y = torch.prod(x)
+
+        return y
+
+    @autotest
```
Comment (BBuf) [inline python/oneflow/test/modules/test_prod.py:20]: 这里写错了，@autotest如果后面不加括号就表示这个样例没有被运行

File: oneflow/core/functional/impl/math_functor.cpp [hunk 3]
```diff
@@ -186,6 +186,30 @@ class ReduceMeanFunctor {
   }
 };
 
+class ReduceProdFunctor {
```
Comment (BBuf) [inline oneflow/core/functional/impl/math_functor.cpp:189]: 你只实现前向，没有实现反向

File: oneflow/core/functional/impl/math_functor.cpp [hunk 4]
```diff
@@ -186,6 +186,30 @@ class ReduceMeanFunctor {
...
+ public:
+  ReduceProdFunctor() {
+    op_ = CHECK_JUST(
+        one::OpBuilder("reduce_prod").Input("input_tensor").Output("output_tensor").Build());
+  }
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& x, const std::vector<int32_t>& axis,
```
Comment (BBuf) [inline oneflow/core/functional/impl/math_functor.cpp:195]: axis改成dim吧

File: oneflow/core/autograd/gradient_funcs/reduce_ops.cpp [hunk 5]
```diff
@@ -64,6 +64,48 @@ Maybe<void> ReduceSumOp::Apply(const ReduceSumOpInterpState* ctx, const TensorTu
...
+  ctx->SaveTensorForBackward(inputs.at(0));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> ReduceProdOp::Apply(const ReduceProdOpInterpState* ctx, const TensorTuple& out_grads,
+                               TensorTuple* in_grads) const {
+  const auto& input = ctx->SavedTensors().at(0);
+  const auto& dy = out_grads.at(0);
+  in_grads->resize(1);
+  in_grads->at(0) = JUST(functional::BroadcastLike(dy, input, ctx->axis));
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/reduce_ops.cpp:103]: 这里求梯度有问题。

File: python/oneflow/test/modules/test_prod.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@flow.unittest.skip_unless_1n1d()
+class TestReduceProd(flow.unittest.TestCase):
+    @autotest()
+    def test_reduce_prod_without_dim(test_case):
+        device = random_device()
+
+        ndim = random(1, 5).to(int)
+        x = random_pytorch_tensor(ndim=ndim).to(device)
+        y = torch.prod(x)
```
Comment (BBuf) [inline python/oneflow/test/modules/test_prod.py:16]: 调整一下格式

File: oneflow/core/autograd/gradient_funcs/reduce_ops.cpp [hunk 7]
```diff
@@ -64,6 +64,53 @@ Maybe<void> ReduceSumOp::Apply(const ReduceSumOpInterpState* ctx, const TensorTu
...
+Maybe<void> ReduceProdOp::Apply(const ReduceProdOpInterpState* ctx, const TensorTuple& out_grads,
+                               TensorTuple* in_grads) const {
+  const auto& input = ctx->SavedTensors().at(0);
+  const auto& output = ctx->SavedTensors().at(1);
+  //const auto& dy = out_grads.at(0);
+
+  const auto& bcast_like = JUST(functional::BroadcastLike(output, input, ctx->axis));
+
+  in_grads->resize(1);
+  in_grads->at(0) = JUST(functional::BroadcastDiv(bcast_like, input));
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/reduce_ops.cpp:108]: 求输入梯度前需要判断requires_grad是否为True再计算，参考其它的反向的样例。

File: oneflow/core/autograd/gradient_funcs/reduce_ops.cpp [hunk 8]
```diff
@@ -64,6 +64,53 @@ Maybe<void> ReduceSumOp::Apply(const ReduceSumOpInterpState* ctx, const TensorTu
...
+  ctx->SaveTensorForBackward(inputs.at(0));
+  ctx->SaveTensorForBackward(outputs.at(0));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> ReduceProdOp::Apply(const ReduceProdOpInterpState* ctx, const TensorTuple& out_grads,
+                               TensorTuple* in_grads) const {
+  const auto& input = ctx->SavedTensors().at(0);
+  const auto& output = ctx->SavedTensors().at(1);
+  //const auto& dy = out_grads.at(0);
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/reduce_ops.cpp:103]: 可以删掉

File: oneflow/core/autograd/gradient_funcs/reduce_ops.cpp (hunk 7 above)
Comment (BBuf) [reply oneflow/core/autograd/gradient_funcs/reduce_ops.cpp:108]: 这里计算了prod op的梯度之后就断了吧，还要乘以dy吧。。。

File: python/oneflow/nn/modules/reduce_ops.py [hunk 9]
```diff
@@ -161,6 +161,44 @@ def _max(input, dim=None, keepdim=False):
     return Max(dim, keepdim)(input)
 
 
+def prod_op(input, dim=None, keepdim=False):
+    r"""Computes the product of row of elements in a tensor in the given axis, if the axis is None, \
```
Comment (BBuf) [inline python/oneflow/nn/modules/reduce_ops.py:165]: 可以补一个note，就是不指定dim的时候dim默认就是Tensor的所有维度，最后获得的Tensor只有一个元素。

File: python/oneflow/test/modules/test_prod.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        return y
+
+    @autotest()
+    def test_reduce_prod_with_dim(test_case):
+        device = random_device()
+
+        ndim = random(1, 5).to(int)
+        x = random_pytorch_tensor(ndim=ndim).to(device)
+
+        dim = random(0, ndim).to(int)
```
Comment (BBuf) [inline python/oneflow/test/modules/test_prod.py:42]: 这里你再接一个y = torch.exp(y) 在`y = torch.prod(x, dim)`后面，来说明你的梯度的改动是work的。

File: python/oneflow/test/modules/test_prod.py (hunk 6 above)
Comment (BBuf) [reply python/oneflow/test/modules/test_prod.py:31]: 去掉空行

File: python/oneflow/nn/modules/reduce_ops.py (hunk 9 above)
Comment (BBuf) [reply python/oneflow/nn/modules/reduce_ops.py:165]: 补note。。

File: oneflow/core/functional/impl/math_functor.cpp (hunk 4 above)
Comment (BBuf) [reply oneflow/core/functional/impl/math_functor.cpp:195]: 这里没改

File: python/oneflow/test/modules/test_prod.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    @autotest()
+    def test_reduce_prod_with_dim(test_case):
+        device = random_device()
+
+        ndim = random(1, 5).to(int)
+        x = random_pytorch_tensor(ndim=ndim).to(device)
+        dim = random(0, ndim).to(int)
+        y = torch.prod(x, dim)
+        y = torch.exp(y)
+
```
Comment (BBuf) [inline python/oneflow/test/modules/test_prod.py:43]: 这个空行也删掉

File: python/oneflow/test/modules/test_prod.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        ndim = random(1, 5).to(int)
+        x = random_pytorch_tensor(ndim=ndim).to(device)
+        y = torch.prod(x)
+
+        return y
+
+    @autotest()
+    def test_reduce_prod_with_dim(test_case):
+        device = random_device()
+
```
Comment (BBuf) [inline python/oneflow/test/modules/test_prod.py:37]: 还有这里

**Review by BBuf — APPROVED** [review]
